File: ADT's/Stack/src/Typescript/LinkedListStack.ts
import { Stack } from './StackADT'

class ListNode<Type>
{
    public data: Type | null;
    public next: ListNode<Type> | null;
    public prev: ListNode<Type> | null;

    constructor(data: Type | null)
    {
        this.data = data;
        this.next = null;
        this.prev = null;
    }

    getNext(): ListNode<Type> | null { return this.next; }
    getPrev(): ListNode<Type> | null { return this.prev; }
    getData(): Type | null { return this.data; }

    setNext(next: ListNode<Type> | null) { this.next = next; }
    setPrev(prev: ListNode<Type> | null) { this.prev = prev; }
}

class LinkedListStack<Type> implements Stack<Type>
{
    private size: number;
    private head: ListNode<Type>;
    private tail: ListNode<Type>;

    constructor()
    {
        this.size = 0;
        this.head = new ListNode<Type>(null);
        this.tail = new ListNode<Type>(null);
        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    push(data: Type)
    {
        const newNode = new ListNode<Type>(data);

        this.tail.prev!.next = newNode;
        newNode.prev = this.tail.prev;

        this.tail.prev = newNode;
        newNode.next = this.tail;

        ++this.size;
    }

    pop() : Type
    {
        if(this.isEmpty()) throw new RangeError("Stack is empty.");
        
        const output = this.tail.prev!.data;
        this.tail.prev = this.tail.prev!.prev;
        this.tail.prev!.next = this.tail;

        --this.size;

        return output!;
    }

    peek() : Type
    {
        if(this.isEmpty()) throw new RangeError("Stack is empty.");
        return this.tail.prev!.data!;
    }

    isEmpty() : boolean
    {
        return this.size === 0;
    }

    toString() : string
    {
        let output = "[";
        let curr = this.head.next!;
        while (curr.next != null)
        {
            output += curr.data + ', ';
            curr = curr.next;
        }
        output = output.slice(0, output.length-2) + ']'
        return output;
    }
}

let stack = new LinkedListStack<number>();
stack.push(5)
stack.push(3)
stack.push(7)
stack.pop()
console.log(stack.peek());
console.log(stack.toString());

export { LinkedListStack };

File: ADT's/Stack/src/Java/LinkedListStack.java
import java.util.EmptyStackException;
import java.util.Iterator;
import java.util.NoSuchElementException;

public class LinkedListStack<E> implements StackADT<E> {

    private static class Node<E> {
        
        public E data;
        public Node<E> next;
        public Node<E> prev;

        public Node(E data) {
            this(data, null, null);
        }
        
        public Node(E data, Node<E> next, Node<E> prev) {
            this.data = data;
            this.next = next;
            this.prev = prev;
        }
    }

    Node<E> head;
    Node<E> tail;        
    int size = 0;

    public LinkedListStack() {
        this.head = new Node<E>(null);
        this.tail = new Node<E>(null);

        this.head.next = this.tail;
        this.tail.prev = this.head;
    }

    
    private void addFirst(E e) {
        linkBefore(this.head.next, e);
    }

    private void linkBefore(Node<E> n, E e) {
        Node<E> newNode = new Node<E>(e);
        
        n.prev.next = newNode;
        newNode.prev = n.prev;
        n.prev = newNode;
        newNode.next = n;
    }

    public void push(E value) {
        addFirst(value);
        size++;
    }

    public E pop() throws EmptyStackException {
        if(isEmpty()) {
            throw new EmptyStackException();
        }

        E value = head.next.data;
        head.next = head.next.next;
        head.next.prev = head;
        size--;
        return value;
    }

    public E peek() throws EmptyStackException {
        if(isEmpty()) {
            throw new EmptyStackException();
        }
        return head.next.data;
    }
    
    public boolean isEmpty() {
        return size == 0;
    }
    
    public int size() {
        return size;
    }

    private class LinkedListStackIterator implements Iterator<E> {
        private Node<E> current;

        public LinkedListStackIterator() {
            current = head;
        }

        public boolean hasNext() {
            return current.next != null;
        }

        public E next() throws NoSuchElementException {
            if(!hasNext()) {
                throw new NoSuchElementException();
            }
            
            current = current.next;
            return current.data;
        }

        public void remove() throws UnsupportedOperationException {
            throw new UnsupportedOperationException();
        }
    }

    public Iterator<E> iterator() {
        return new LinkedListStackIterator();
    }

}
